programs/stake/src/stake_instruction.rs

```diff
@@ -5969,4 +5995,137 @@ mod tests {
+                is_signer: false,
+                is_writable: false,
+            },
+            AccountMeta {
+                pubkey: rent_address,
+                is_signer: false,
+                is_writable: false,
+            },
+        ];
+        let good_accounts = (good_transaction_accounts, good_instruction_accounts);
+
+        // The instruction data needs to deserialize to a bogus StakeInstruction.  We likely never
+        // will have `usize::MAX`-number of instructions, so this should be a safe constant to
+        // always map to an invalid stake instruction.
+        let bad_instruction = Instruction::new_with_bincode(id(), &usize::MAX, Vec::default());
+        let bad_transaction_accounts = Vec::default();
+        let bad_instruction_accounts = Vec::default();
+        let bad_accounts = (bad_transaction_accounts, bad_instruction_accounts);
+
+        for (
+            is_feature_enabled,
+            instruction,
+            (transaction_accounts, instruction_accounts),
+            expected_result,
+        ) in [
+            (true, &good_instruction, &good_accounts, Ok(())),
+            (
+                true,
+                &bad_instruction,
+                &good_accounts,
+                Err(InstructionError::InvalidInstructionData),
+            ),
+            (
+                true,
+                &good_instruction,
+                &bad_accounts,
+                Err(InstructionError::NotEnoughAccountKeys),
+            ),
+            (
+                true,
+                &bad_instruction,
+                &bad_accounts,
+                Err(InstructionError::InvalidInstructionData),
+            ),
+            (false, &good_instruction, &good_accounts, Ok(())),
+            (
+                false,
+                &bad_instruction,
+                &good_accounts,
+                Err(InstructionError::InvalidInstructionData),
+            ),
+            (
+                false,
+                &good_instruction,
+                &bad_accounts,
+                Err(InstructionError::NotEnoughAccountKeys),
+            ),
+            (
+                false,
+                &bad_instruction,
+                &bad_accounts,
+                Err(InstructionError::NotEnoughAccountKeys),
+            ),
+        ] {
+            mock_process_instruction(
+                &id(),
+                Vec::new(),
+                &instruction.data,
+                transaction_accounts.clone(),
+                instruction_accounts.clone(),
+                None,
+                expected_result,
+                if is_feature_enabled {
+                    |first_instruction_account, invoke_context| {
+                        super::process_instruction(first_instruction_account, invoke_context)
+                    }
+                } else {
+                    |first_instruction_account, invoke_context| {
+                        let mut feature_set = FeatureSet::all_enabled();
```

Comment:
The feature set could be created outside of the closure / lambdas here, somewhat simplifying the `if` blocks.